### DISCUSSION TITLE
-define connected-components question

### DISCUSSION THREAD

**Discussion by bioworldmerch** [ImageMagick/ImageMagick#5163] [Digital Image Processing]
Fred or Anthony I'm sure this ones for you....about the -define connected-components commands:

Can someone explain to me what these three parts of this command do independently and the effects of changing them?:
Magick imagein.png -define connected-components:area-threshold=20 -define connected-components:mean-color=true -connected-components 8 imageout.png

The whole command does a nice job at finding floating pixels of color and applying nearest color on some images but I'm having issues with error of 'Magick: too many objects error/vision.c/ConnectedComponentsImage/1091' on more complicated images. I would like to find a way to either raise the tolerance up enough to find less components or handle more then the limit of components or apply to the found components instead of having no effect after the error. Another solution would be also welcome even if it is less effective.

I seem to remember at one point Fred having a way of dividing an image into quadrants that I could apply the command to each independently but couldn't find it.

I also tried adding a morphology erodeIntensity or closeIntensity first to the image without losing too much detail.

Thank very much!

**Comment by fmw42** [comment]
`-define connected-components:area-threshold=20  `

This removes any regions that are smaller than 20 pixels in area. They get merged into the surrounding regions.

`-define connected-components:mean-color=true `

This cause the output image to present the regions by their mean (average) color rather than by their ID number

`-connected-components 8`

This is the required operator for connected-components and sets the type of connection to define regions to 8 neighbors rather than 4 neighbors.

The maximum number of regions permitted in an output is determined by the compile Q level. For Q8 or 8-bits you can have 256 regions max. For Q16 or 16-bits your can 65535 regions max. For Q32, you can have millions of regions max.  The issue is that connected components does not work well will true color images with 24-bit color. You should either threshold or reduce colors to limit the number of regions.

Please characterize your images! And you objectives.

I do not recall anything that I created that divided the image into quadrants.  You could do that, but then the regions would be split along the quadrant boundaries compared to the full image.

**Reply by bioworldmerch** [reply]
Im reducing colors via kmeans for images prior to running this and using it to remove much of the intermediate colors due to aliasing. It's working very well for images that aren't too complicated. I figured it is too many points or parts to the image and not related to 24 bit or the colors. It's pretty detailed. Im trying to find a scenario that will work either way or have a plan b when the image can't be processed.

**Reply by fmw42** [reply]
It does depend upon the number of colors you use to reduce to in -kmeans. The more colors you have, the more regions you get.

**Comment by snibgo** [comment]
From a quick look at vision.c, the "too many objects" error depends on MaxColormapSize, a constant that is #defined in magick-type.h, and has a maximum value of 65536. In vision.c, this determines the size of "object" array.

I suspect that vision.c could examine (for example) "-define MaxColormapSize" and, if that is set, use that value instead of the constant MaxColormapSize.

**Reply by bioworldmerch** [reply]
That would be great. Its beyond my capabilities to make the change but it would improve the connected components functionality greatly.

**Reply by bioworldmerch** [reply]
@snibgo after having unpredictable errors of too many objects for define connected components, I am realizing it is because it errors out based on 65536 which is probably way too low when I am using 300 dpi images. Can you suggest how I can increase that number and where to find the max value and replace with a much higher number as test?

**Comment by snibgo** [comment]
As a test, I suggest inserting ...
```
#define MaxColormapSize  1000000UL
```
... or some other number somewhere near the top of vision.c. Then rebuild IM. The compiler may warn you that MaxColormapSize has been redefined.

**Reply by bioworldmerch** [reply]
I've been trying to install via compiler on Mac and keep getting a stupid error. 

./configure works fine but make gives me

ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [utilities/magick] Error 1
make: *** [all] Error 2

It installs fine via Brew so Im going to try editing the Cached folder there.

**Reply by bioworldmerch** [reply]
@snibgo  I was able to reinstall In brew after editing but when I run the command 

`Magick imagein.png -define connected-components:area-threshold=20 -define connected-components:mean-color=true -connected-components 8 imageout.png`

I get the following errors:

`magick: memory allocation failed `/Horse.png' @ error/colormap.c/AcquireImageColormap/123.
magick: memory allocation failed `Horse.png' @ error/vision.c/ConnectedComponentsImage/860.`

**Reply by fmw42** [reply]
Perhaps you have too many colors and thus too many regions. Try thresholding your image first or reducing the number of colors. Post your horse.png image so we can see what you are using.

Note on Q16 compile, you can only have 65535 regions. On Q8, only 256 regions.

**Reply by bioworldmerch** [reply]
I tried Q32 too before editing the MaxColormapSize in vision.c snibgo suggested.

![Horse_proof](https://user-images.githubusercontent.com/84971106/171966930-71342cc2-e667-4730-935b-a1abfdc66db0.png)

**Reply by bioworldmerch** [reply]
@fmw42 also tried reducing it to 256 colors indexed then convert back to rgb and still errors. I believe its the dimension which is 3535 x 2357 and color count impacting it over the 65536k limit in Q16. I sw no difference in the Q32.

**Comment by fmw42** [comment]
It works for me reducing colors to 4, which what I see in your image. It took about 10 sec on my Mac mini.  You can try more colors if you want.



```
magick horse.png -kmeans 4 \
-define connected-components:verbose=true \
-define connected-components:area-threshold=20 \
-define connected-components:mean-color=true \
-connected-components 8 horse_ccl.png

Objects (id: bounding-box centroid area mean-color):
  0: 3600x4200+0+0 1798.5,2100.7 1.34902e+07 srgba(99.9986%,82.7455%,56.8651%,1.00486)
  38: 1182x1392+1217+1393 1802.2,2066.2 710614 srgba(1.87885%,37.3233%,50.0369%,1.09222)
  4: 1329x1558+1144+1318 1808.2,2219.7 217354 srgba(1.87885%,37.3233%,50.0369%,1.30151)
  19: 1256x1478+1180+1354 1807.6,2161.0 188277 srgba(68.0982%,72.4798%,73.9828%,1.34808)
  99: 737x1062+1438+1542 1819.0,2040.8 182284 srgba(98.5009%,99.0753%,99.1188%,1.35952)
  111: 432x395+1840+1610 2070.7,1810.5 84490 srgba(68.0982%,72.4798%,73.9828%,1.77565)
  103: 304x208+1529+1549 1705.2,1661.5 32756 srgba(98.5009%,99.0753%,99.1188%,3.0007)
  95: 744x1070+1434+1539 1818.6,2042.5 22121 srgba(68.0982%,72.4798%,73.9828%,3.96257)
  2: 1336x1563+1141+1316 1810.8,2193.3 20064 srgba(68.0982%,72.4798%,73.9828%,4.29636)
  168: 195x164+1487+2116 1597.4,2214.5 14179 srgba(68.0982%,72.4798%,73.9828%,5.62198)
  195: 262x153+1602+2320 1735.1,2384.2 14138 srgba(98.5009%,99.0753%,99.1188%,5.63538)
  174: 141x189+1652+2125 1739.1,2223.9 13887 srgba(98.5009%,99.0753%,99.1188%,5.71916)
  133: 211x213+1387+1809 1482.2,1894.8 13406 srgba(98.5009%,99.0753%,99.1188%,5.88848)
  205: 113x218+1700+2397 1753.7,2515.8 7458 srgba(68.0982%,72.4798%,73.9828%,9.78721)
  149: 155x124+2072+1954 2160.3,2019.9 6646 srgba(68.0982%,72.4798%,73.9828%,10.8608)
  108: 91x187+1939+1588 1980.0,1664.6 5266 srgba(68.0982%,72.4798%,73.9828%,13.4449)
  115: 211x51+1352+1649 1465.9,1675.0 4885 srgba(68.0982%,72.4798%,73.9828%,14.4156)
  159: 102x119+1508+2047 1553.0,2113.1 4864 srgba(68.0982%,72.4798%,73.9828%,14.4735)
  101: 309x214+1527+1546 1669.9,1678.3 4600 srgba(68.0982%,72.4798%,73.9828%,15.2467)
  117: 121x118+1348+1697 1392.8,1735.6 4436 srgba(68.0982%,72.4798%,73.9828%,15.7734)
  126: 120x105+1378+1765 1440.7,1817.7 3829 srgba(68.0982%,72.4798%,73.9828%,18.1154)
  156: 122x50+1952+2015 2009.2,2038.1 3735 srgba(68.0982%,72.4798%,73.9828%,18.5462)
  185: 142x85+1533+2244 1600.3,2296.9 3671 srgba(68.0982%,72.4798%,73.9828%,18.8521)
  121: 166x123+1376+1718 1446.9,1760.0 3597 srgba(68.0982%,72.4798%,73.9828%,19.2194)
  193: 269x166+1599+2314 1753.7,2393.0 3177 srgba(68.0982%,72.4798%,73.9828%,21.628)
  198: 100x68+1575+2334 1610.5,2358.2 3041 srgba(68.0982%,72.4798%,73.9828%,22.5505)
  253: 83x64+1545+2659 1588.8,2690.3 2724 srgba(98.5009%,99.0753%,99.1188%,25.0584)
  257: 83x64+2081+2659 2124.9,2690.5 2704 srgba(98.5009%,99.0753%,99.1188%,25.2363)
  132: 220x219+1382+1806 1483.4,1901.2 2568 srgba(68.0982%,72.4798%,73.9828%,26.5199)
  245: 161x71+2006+2655 2088.5,2694.5 2521 srgba(68.0982%,72.4798%,73.9828%,26.9956)
  172: 152x197+1647+2122 1730.0,2239.0 2316 srgba(68.0982%,72.4798%,73.9828%,29.2966)
  208: 62x89+1655+2420 1690.7,2473.7 2293 srgba(68.0982%,72.4798%,73.9828%,29.5805)
  242: 135x71+1694+2655 1765.2,2692.9 2204 srgba(68.0982%,72.4798%,73.9828%,30.7346)
  141: 73x57+1463+1897 1492.5,1922.7 2150 srgba(68.0982%,72.4798%,73.9828%,31.4814)
  252: 73x65+1458+2659 1494.2,2691.2 2139 srgba(98.5009%,99.0753%,99.1188%,31.6381)
  209: 37x88+1805+2426 1822.9,2468.1 2059 srgba(68.0982%,72.4798%,73.9828%,32.8286)
  256: 60x64+1937+2659 1966.4,2690.7 2005 srgba(98.5009%,99.0753%,99.1188%,33.6858)
  254: 60x64+1766+2659 1795.6,2690.4 1976 srgba(98.5009%,99.0753%,99.1188%,34.1655)
  105: 81x70+1885+1561 1933.4,1584.8 1926 srgba(68.0982%,72.4798%,73.9828%,35.0265)
  248: 55x65+1643+2658 1663.3,2685.5 1856 srgba(98.5009%,99.0753%,99.1188%,36.3098)
  202: 32x114+1953+2377 1969.9,2431.7 1852 srgba(68.0982%,72.4798%,73.9828%,36.3861)
  249: 60x65+1697+2658 1729.9,2693.9 1617 srgba(98.5009%,99.0753%,99.1188%,41.5288)
  250: 59x65+2010+2658 2042.2,2693.8 1611 srgba(98.5009%,99.0753%,99.1188%,41.6797)
  251: 43x65+2179+2658 2194.7,2690.4 1575 srgba(98.5009%,99.0753%,99.1188%,42.6095)
  240: 90x71+1541+2655 1585.6,2692.5 1499 srgba(68.0982%,72.4798%,73.9828%,44.7191)
  247: 56x67+1394+2657 1413.8,2690.4 1411 srgba(98.5009%,99.0753%,99.1188%,47.4458)
  255: 51x64+1842+2659 1865.3,2687.4 1372 srgba(98.5009%,99.0753%,99.1188%,48.766)
  239: 80x73+1455+2654 1494.3,2688.3 1298 srgba(68.0982%,72.4798%,73.9828%,51.4892)
  258: 29x50+1480+2666 1494.1,2690.2 1228 srgba(1.87885%,37.3233%,50.0369%,54.3673)
  244: 66x71+1934+2655 1967.1,2688.8 1216 srgba(68.0982%,72.4798%,73.9828%,54.8939)
  124: 105x72+1771+1753 1827.5,1795.8 1141 srgba(1.87885%,37.3233%,50.0369%,58.4365)
  122: 129x89+1754+1742 1822.0,1795.5 1103 srgba(68.0982%,72.4798%,73.9828%,60.4152)
  246: 49x71+2176+2655 2198.4,2689.6 1052 srgba(68.0982%,72.4798%,73.9828%,63.2956)
  241: 60x71+1640+2655 1663.8,2682.0 988 srgba(68.0982%,72.4798%,73.9828%,67.331)
  238: 63x73+1390+2654 1422.3,2689.9 959 srgba(68.0982%,72.4798%,73.9828%,69.3368)
  243: 60x71+1837+2655 1867.7,2682.5 857 srgba(68.0982%,72.4798%,73.9828%,77.4702)
  88: 54x37+1751+1530 1773.3,1546.6 789 srgba(98.5009%,99.0753%,99.1188%,84.0608)
  188: 27x92+2019+2273 2034.1,2317.8 698 srgba(68.0982%,72.4798%,73.9828%,94.8897)
  190: 17x71+2023+2283 2031.8,2316.6 639 srgba(98.5009%,99.0753%,99.1188%,103.559)
  85: 69x44+1747+1527 1779.7,1543.3 563 srgba(68.0982%,72.4798%,73.9828%,117.403)
  131: 39x22+1565+1794 1585.8,1803.7 450 srgba(98.5009%,99.0753%,99.1188%,146.633)
  130: 55x32+1553+1788 1581.5,1800.7 445 srgba(68.0982%,72.4798%,73.9828%,148.27)
  110: 27x16+1677+1600 1689.0,1606.1 317 srgba(1.87885%,37.3233%,50.0369%,207.735)
  109: 32x23+1675+1597 1689.6,1606.1 238 srgba(68.0982%,72.4798%,73.9828%,276.357)
  259: 11x20+1665+2667 1669.3,2676.2 183 srgba(1.87885%,37.3233%,50.0369%,359.115)
  260: 10x13+1717+2688 1721.4,2694.7 103 srgba(1.87885%,37.3233%,50.0369%,637.262)
  261: 10x12+2029+2689 2033.4,2695.1 97 srgba(1.87885%,37.3233%,50.0369%,676.619)
  266: 88x1+2347+2728 2390.5,2728.0 88 srgba(98.5009%,99.0753%,99.1188%,745.716)
  265: 77x1+1193+2728 1231.0,2728.0 77 srgba(98.5009%,99.0753%,99.1188%,852.104)
  226: 38x1+2351+2535 2369.5,2535.0 38 srgba(98.5009%,99.0753%,99.1188%,1725.61)
  268: 1x27+1271+2747 1271.0,2760.0 27 srgba(98.5009%,99.0753%,99.1188%,2428.22)
  113: 2x20+1145+1618 1145.4,1627.7 22 srgba(98.5009%,99.0753%,99.1188%,2979.86)
```


![horse_ccl](https://user-images.githubusercontent.com/13458218/171967999-e89745a5-8aa1-4e6d-a66b-8af36f19cd2c.png)

**Reply by bioworldmerch** [reply]
@fmw42 that did work for me too. the issue is sometimes converting down colors via kmeans grabs the wrong color. It would be better not to hit the ceiling of 65k but I think maybe I can try and figure out what the cap is for color count that still holds the important colors. The cap for this image is somewhere between 212 colors and what you did for 4.

**Comment by snibgo** [comment]
```
magick: memory allocation failed /Horse.png' @ error/colormap.c
```
So, did you try other numbers between 65000 and 1000000?

**Reply by bioworldmerch** [reply]
@snibgo I will now. I tried going up. What's 1000000UL vs 1000000?

**Comment by snibgo** [comment]
"UL" is unsigned long. I suggest UL as that is what magick-type.h has.

Your error is in colormap.c, probably from:
```
  if (colors > MaxColormapSize)
```
... so you might try inserting that define also at the top of that file.

I stress that these are just hacks to see if it works, not a proper solution.

**Reply by bioworldmerch** [reply]
completely understand
